packages/panzer/disc-fe/src/Panzer_BasisValues2_impl.hpp

```diff
@@ -45,1711 +45,1628 @@
+  }
+  else { TEUCHOS_ASSERT(false); }
+
+  basis_coordinates_ref = af.buildStaticArray<Scalar,BASIS,Dim>("basis_coordinates_ref",card,dim);
+  basis_coordinates = af.buildStaticArray<Scalar,Cell,BASIS,Dim>("basis_coordinates",numcells,card,dim);
+}
+
+
+//=======================================================================================================
+// New Interface
+
+
+template <typename Scalar>
+void
+BasisValues2<Scalar>::
+setup(const Teuchos::RCP<const panzer::BasisIRLayout> & basis,
+      PHX::MDField<Scalar, Cell, IP, Dim>               reference_points,
+      PHX::MDField<Scalar, Cell, IP, Dim, Dim>          point_jacobian,
+      PHX::MDField<Scalar, Cell, IP>                    point_jacobian_determinant,
+      PHX::MDField<Scalar, Cell, IP, Dim, Dim>          point_jacobian_inverse,
+      const Teuchos::RCP<const OrientationsInterface> & orientations,
+      const std::string &                               prefix_,
+      const std::vector<PHX::index_size_type> &         ddims)
+{
+  basis_layout = basis;
+  intrepid_basis = basis->getBasis()->getIntrepid2Basis<PHX::Device::execution_space,Scalar,Scalar>();
+  num_cells_ = basis_layout->numCells();
+  num_evaluate_cells_ = num_cells_;
+  num_real_cells_ = num_cells_;
+  prefix = prefix_;
+  ddims_ = ddims;
+  orientations_ = orientations;
+
+  cubature_points_ref_ = reference_points;
+  cubature_jacobian_ = point_jacobian;
+  cubature_jacobian_determinant_ = point_jacobian_determinant;
+  cubature_jacobian_inverse_ = point_jacobian_inverse;
+
+}
+
+template <typename Scalar>
+void
+BasisValues2<Scalar>::
+setupUniform(const Teuchos::RCP<const panzer::BasisIRLayout> &  basis,
+             PHX::MDField<Scalar, IP, Dim>                      reference_points,
+             PHX::MDField<Scalar, Cell, IP, Dim, Dim>           point_jacobian,
+             PHX::MDField<Scalar, Cell, IP>                     point_jacobian_determinant,
+             PHX::MDField<Scalar, Cell, IP, Dim, Dim>           point_jacobian_inverse,
+             const Teuchos::RCP<const OrientationsInterface> &  orientations,
+             const std::string &                                prefix_,
+             const std::vector<PHX::index_size_type> &          ddims)
+{
+  basis_layout = basis;
+  intrepid_basis = basis->getBasis()->getIntrepid2Basis<PHX::Device::execution_space,Scalar,Scalar>();
+  num_cells_ = basis_layout->numCells();
+  num_evaluate_cells_ = num_cells_;
+  num_real_cells_ = num_cells_;
+  prefix = prefix_;
+  ddims_ = ddims;
+  cubature_points_uniform_ref_ = reference_points;
+  orientations_ = orientations;
+
+  cubature_jacobian_ = point_jacobian;
+  cubature_jacobian_determinant_ = point_jacobian_determinant;
+  cubature_jacobian_inverse_ = point_jacobian_inverse;
+}
+
+template <typename Scalar>
+void
+BasisValues2<Scalar>::
+setWeightedMeasure(PHX::MDField<Scalar, Cell, IP> weighted_measure)
+{
+  cubature_weights_ = weighted_measure;
```

Comment:
Yeah, this is a challenging one. I'm not exactly sure how allow the geometry to change such that this object "automatically updates" - it is hard to keep track without a bunch of complicated bools. I added a "force" bool on all the request commands to enable updating, but it requires the user to know what they are doing. I like your idea of adding a throw here if weights have already been set. Each time we call "setup" it will "clear" the arrays that reset the weights.

Reply:
I made some changes, but it requires some strange order of operations during construction. It can be cleaned up once we remove the deprecated path. I added documentation to the class description to describe the correct construction path.